File: easy-projects/easy-common/src/main/java/xyz/srclab/common/util/proxy/ProxyClass.java
package xyz.srclab.common.util.proxy;

import xyz.srclab.annotation.Immutable;

import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.function.Predicate;

@Immutable
public interface ProxyClass<T> {

    static <T> Builder<T> newBuilder(Class<T> type) {
        return ProxyClassSupport.newBuilder(type);
    }

    T newInstance();

    T newInstance(Class<?>[] parameterTypes, Object[] args);

    Class<T> getProxyClass();

    interface Builder<T> {

        default Builder<T> proxyMethod(String methodName, Class<?>[] parameterTypes, ProxyMethod proxyMethod) {
            return proxyMethod(
                    method -> method.getName().equals(methodName)
                            && Arrays.equals(method.getParameterTypes(), parameterTypes),
                    proxyMethod
            );
        }

        Builder<T> proxyMethod(Predicate<Method> methodPredicate, ProxyMethod proxyMethod);

        ProxyClass<T> build();
    }
}

File: srclab-common/src/main/kotlin/xyz/srclab/common/proxy/ProxyClass.kt
package xyz.srclab.common.proxy

import xyz.srclab.common.base.CachingProductBuilder
import xyz.srclab.common.base.asNotNull
import java.util.*

/**
 * @author sunqian
 */
interface ProxyClass<T : Any> {

    fun newInstance(): T

    fun newInstance(parameterTypes: Array<Class<*>>?, args: Array<Any?>?): T

    //fun getProxyClass(): Class<T>

    class Builder<T : Any>(private val proxyClass: Class<T>) : CachingProductBuilder<ProxyClass<T>>() {

        private var proxyMethods: MutableList<ProxyMethod<T>>? = null

        fun addProxyMethod(proxyMethod: ProxyMethod<T>) = apply {
            notNullProxyMethods().add(proxyMethod)
            commitChange()
        }

        fun addProxyMethods(vararg proxyMethods: ProxyMethod<T>) = apply {
            notNullProxyMethods().addAll(proxyMethods)
            commitChange()
        }

        fun addProxyMethods(proxyMethods: Iterable<ProxyMethod<T>>) = apply {
            notNullProxyMethods().addAll(proxyMethods)
            commitChange()
        }

        override fun buildNew(): ProxyClass<T> {
            return ProxyClassGenerator.DEFAULT.generate(proxyClass, proxyMethods())
        }

        private fun notNullProxyMethods(): MutableList<ProxyMethod<T>> {
            if (proxyMethods === null) {
                proxyMethods = LinkedList()
            }
            return proxyMethods.asNotNull()
        }

        private fun proxyMethods(): List<ProxyMethod<T>> {
            val hs = proxyMethods
            if (hs === null) {
                return emptyList()
            }
            return hs.toList()
        }
    }
}